src/main/java/me/botsko/prism/actions/BlockAction.java

```diff
@@ -22,582 +20,546 @@
+        // @todo if we ever track rollback/restore for undo, we'll
+        // need logic to do the opposite
+        return placeBlock(player, parameters, isPreview, block, false);
+
+    }
+
+    @Override
+    public ChangeResult applyDeferred(Player player, QueryParameters parameters, boolean isPreview) {
+        final Block block = getWorld().getBlockAt(getLoc());
+        return placeBlock(player, parameters, isPreview, block, true);
+    }
+
+    ChangeResult placeBlock(Player player, QueryParameters parameters, boolean isPreview, Block block,
+                            boolean isDeferred) {
+        BlockStateChange stateChange;
+        BlockState state = block.getState();
+
+        // Ensure block action is allowed to place a block here.
+        // (essentially liquid/air).
+
+        final boolean cancelIfBadPlace = !getActionType().requiresHandler(BlockChangeAction.class)
+                && !getActionType().requiresHandler(PrismRollbackAction.class) && !parameters.hasFlag(Flag.OVERWRITE);
+
+        if (cancelIfBadPlace && !BlockUtils.isAcceptableForBlockPlace(block.getType())) {
+            // System.out.print("Block skipped due to being unacceptable for block place.");
+            return new ChangeResult(ChangeResultType.SKIPPED, null);
+        }
+
+        // On the blacklist (except an undo)
+        if (Prism.getIllegalBlocks().contains(getMaterial())
+                && !parameters.getProcessType().equals(PrismProcessType.UNDO)) {
+            // System.out.print("Block skipped because it's not allowed to be placed.");
+            return new ChangeResult(ChangeResultType.SKIPPED, null);
+        }
+
+        // If we're not in a preview, actually apply this block
+        // Capture the block before we change it
+        final BlockState originalBlock = block.getState();
+        if (!isPreview) {
+
+            // Capture the block before we change it
+
+            // If lilypad, check that block below is water. Be sure
+            // it's set to stationary water so the lilypad will sit
+            if (getMaterial() == LILY_PAD) {
+
+                final Block below = block.getRelative(BlockFace.DOWN);
+                if (below.getType().equals(WATER) || below.getType().equals(AIR)) {
+                    below.setType(WATER);
+                } else {
+                    // Prism.debug("Lilypad skipped because no water exists below.");
+                    return new ChangeResult(ChangeResultType.SKIPPED, null);
+                }
+            }
+
+            // If portal, we need to light the portal. seems to be the only way.
+            if (getMaterial() == NETHER_PORTAL) {
+                final Block obsidian = BlockUtils.getFirstBlockOfMaterialBelow(OBSIDIAN, block.getLocation());
+                if (obsidian != null) {
+                    final Block above = obsidian.getRelative(BlockFace.UP);
+                    if (!(above.getType() == NETHER_PORTAL)) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use opposite operator instead of the logic complement operator.](https://app.codacy.com/gh/AddstarMC/Prism-Bukkit/pullRequest?prid=5438808)